src/compiler/checker.ts

```diff
@@ -3993,6 +4027,37 @@ namespace ts {
             return links.resolvedType;
         }
 
+        // We do not perform supertype reduction on intersection types. Intersection types are created only by the &
+        // type operator and we can't reduce those because we want to support recursive intersection types. For example,
+        // a type alias of the form "type List<T> = T & { next: List<T> }" cannot be reduced during its declaration.

```

Comment:
I think this argument applies for union types as well. Maybe we shouldn't do subtype reduction for union types either. Many people try to make recursive union types, and we give an error because a union type depends on its constituents. Can we consider abandoning subtype reduction?